### PR TITLE
Fix final-vars type inferring

### LINKED ISSUE #18166: [Spec Deviation] Inferred type for final variables declared with var should be the precise type
**Description:**
According to the spec:

**Local variable declaration statements**
If the typed-binding-pattern uses var, then the type of the variable is inferred from the static type of the action-or-expr; if the local-init-var-decl-stmt includes final, the precise type is used, and otherwise the broad type is used.

**Module variable declaration**
If the typed-binding-pattern uses var, then the type of the variable is inferred from the static type of expression; if the module-var-decl includes final, the precise type is used, and otherwise the broad type is used.

But the following doesn't work.

**Steps to reproduce:**
```ballerina
const ONE = 1;

public function main() {
    final var v1 = ONE;
    ONE v2 = v1; // incompatible types: expected '1', found 'int'
}
```

**Affected Versions:**
```cmd
$ ballerina -v
Ballerina 1.0.0-alpha3
Language specification 2019R2
Ballerina tool 1.0.0
```

### PR DESCRIPTION
**Pull request by irshadnilam**
## Purpose
$title
Fixes #18166

## Remarks
Came across https://github.com/ballerina-platform/ballerina-lang/issues/22959 while fixing this. This can be improved for const maps once it is fixed. 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/var/VarFinalInferredNegativeTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.testng.AssertJUnit.assertEquals;
+
+/**
+ * Class to test final var declarations type inferring.
+ *
+ * @since 3.0.0
+ */
```
Comment (irshadnilam) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/var/VarFinalInferredNegativeTest.java:33]: This should be 2.0, :)
will fix that with other review comments once it is reviewed.

**Comment by KRVPerera** [comment]
May be not related to this PR. But I notice that a error message which may deceive the user.
#### Code
```Ballerina
    final var x = 1.6;
    x = 4.5;
```

#### Output
```LogTalk
error: .::t1.bal:13:5: cannot assign a value to final 'x'
error: .::t1.bal:13:9: incompatible types: expected '1.6f', found 'float'
```

I think giving the second error message on the same line is wrong after the previous error. Probably due to `checkConstantAssignment` not setting expression type to semantic error.

https://github.com/ballerina-platform/ballerina-lang/blob/3831af1e92563161e20d1f33023586f52ce0f43f/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java#L2793

https://github.com/ballerina-platform/ballerina-lang/blob/3831af1e92563161e20d1f33023586f52ce0f43f/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java#L2139

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -982,6 +989,30 @@ void handleDeclaredVarInForeach(BLangVariable variable, BType rhsType, SymbolEnv
...
+            case SIMPLE_VARIABLE_REF:
+                BLangSimpleVarRef ref = (BLangSimpleVarRef) expr;
+                if (ref.symbol.getKind() == SymbolKind.CONSTANT) {
+                    return ((BConstantSymbol) ref.symbol).type;
+                }
+                return originalType;
+            case NUMERIC_LITERAL:
+            case LITERAL:
+                return new BFiniteType(createTypeSymbol(SymTag.FINITE_TYPE), Collections.singleton(expr));
+            case LIST_CONSTRUCTOR_EXPR:
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1004]: Can we mention in https://github.com/ballerina-platform/ballerina-lang/issues/22959 that we need to add support for mapping constructor expressions here too?

**Comment by irshadnilam** [comment]
Closing this temporarily.

**Pull request closed without merging**